Test Failure: SessionCacheOneServerTest.testConcurrentSetGetAndRemove: testConcurrentSetGetAndRemove-key not found

Test Failure: com.ibm.ws.session.cache.fat.SessionCacheOneServerTest.testConcurrentSetGetAndRemove

This is an intermittent failure in the com.ibm.ws.session.cache_fat bucket.

junit.framework.AssertionFailedError: 2018-05-03-23:11:46:428 Servlet call was not successful: ERROR: Caught exception attempting to call test method testSessionInfoCache on servlet session.cache.web.SessionCacheTestServlet java.lang.AssertionError: [testConcurrentSetGetAndRemove-key] not found in [] at session.cache.web.SessionCacheTestServlet.testSessionInfoCache(SessionCacheTestServlet.java:570) at componenttest.app.FATServlet.doGet(FATServlet.java:63) at javax.servlet.http.HttpServlet.service(HttpServlet.java:687) at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255) at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743) at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440) at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1154) at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1004) at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75) at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:927) at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279) at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1011) at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:414) at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:373) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:530) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:464) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:329) at com.ibm.ws.http.channel.internal.inbound.HttpICLReadCallback.complete(HttpICLReadCallback.java:70) at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:501) at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:571) at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:926) at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1015) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) at java.lang.Thread.run(Thread.java:785) at com.ibm.ws.session.cache.fat.FATSuite.run(FATSuite.java:51) at com.ibm.ws.session.cache.fat.SessionCacheApp.invokeServlet(SessionCacheApp.java:29) at com.ibm.ws.session.cache.fat.SessionCacheOneServerTest.testConcurrentSetGetAndRemove(SessionCacheOneServerTest.java:287)

--------------------------------------------
Error in server log:
--------------------------------------------
AssertionError: [testConcurrentSetGetAndRemove-key] not found in []
[5/3/18 23:11:46:409 UTC] 00000041 SystemErr                                                    R java.lang.AssertionError: [testConcurrentSetGetAndRemove-key] not found in []
    at org.junit.Assert.fail(Assert.java:93)
    at org.junit.Assert.assertTrue(Assert.java:43)
    at session.cache.web.SessionCacheTestServlet.testSessionInfoCache(SessionCacheTestServlet.java:570)


The problem is caused by sessionPostInvoke processing still going on after control is returned to the test case.  See the time gap here:

[5/3/18 23:11:46:331 UTC] 00000023 id=         SessionContext                                               > sessionPostInvoke ENTRY
....
[5/3/18 23:11:46:371 UTC] 00000023 id=         SessionManager                                               > releaseSession ENTRY sessionID = PJyZjG-n8uTHI4-vYJ8tke1

During this time, we see the test proceed to read a value,

[5/3/18 23:11:46:366 UTC] 00000041 id=         SystemOut                                                    O BEGIN http://localhost:8010/sessionCacheApp/SessionCacheTestServlet?testMethod=testAttributeIsAnyOf&type=java.lang.Integer&key=testConcurrentSetGetAndRemove-key&values=null,1000,1002,1004,1006,1008,1010
[5/3/18 23:11:46:367 UTC] 00000041 id=         BackedSession                                                > getValueGuts ENTRY
[5/3/18 23:11:46:367 UTC] 00000041 id=         BackedSession                                                1 getValueGuts Value from non-swappable is null for id testConcurrentSetGetAndRemove-key
[5/3/18 23:11:46:367 UTC] 00000041 id=         BackedSession                                                1 getValueGuts Retrieved cached object for testConcurrentSetGetAndRemove-key
[5/3/18 23:11:46:367 UTC] 00000041 id=         BackedSession                                                < getValueGuts RETURN
[5/3/18 23:11:46:367 UTC] 00000041 id=         BackedSession                                                < getAttribute RETURN value= 1010
[5/3/18 23:11:46:367 UTC] 00000041 id=         SystemOut                                                    O Got entry: testConcurrentSetGetAndRemove-key=1010 from sessionID=PJyZjG-n8uTHI4-vYJ8tke1
[5/3/18 23:11:46:368 UTC] 00000041 id=         SystemOut                                                    O END

Having read the above value, it believes that it should find the value in the cache.  However, before it checks the cache, thread 00000023 regains control and removes the cache entry, causing the test to not find it.

We can avoid the issue by forcing a sync of the session prior to the servlet ending.

--- a/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/SessionCacheOneServerTest.java
+++ b/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/SessionCacheOneServerTest.java
@@ -81,7 +81,7 @@ public class SessionCacheOneServerTest extends FATServletClient {
             puts.add(new Callable<Void>() {
                 @Override
                 public Void call() throws Exception {
-                    app.sessionPut("testConcurrentPutNewAttributesAndRemove-key" + offset, (char) ('A' + offset), session, true);
+                    app.sessionPut("testConcurrentPutNewAttributesAndRemove-key" + offset + "&sync=true", (char) ('A' + offset), session, true);
                     return null;
                 }
             });
@@ -96,13 +96,13 @@ public class SessionCacheOneServerTest extends FATServletClient {
                 removes.add(new Callable<Void>() {
                     @Override
                     public Void call() throws Exception {
-                        app.invokeServlet("sessionRemoveAttribute&key=testConcurrentPutNewAttributesAndRemove-key" + offset, session);
+                        app.invokeServlet("sessionRemoveAttribute&key=testConcurrentPutNewAttributesAndRemove-key" + offset + "&sync=true", session);
                         return null;
                     }
                 });
         }
 
-        String sessionId = app.sessionPut("testConcurrentPutNewAttributesAndRemove-key0", 'A', session, true);
+        String sessionId = app.sessionPut("testConcurrentPutNewAttributesAndRemove-key0&sync=true", 'A', session, true);
         try {
             List<Future<Void>> futures = executor.invokeAll(puts);
             for (Future<Void> future : futures)
@@ -166,7 +166,7 @@ public class SessionCacheOneServerTest extends FATServletClient {
                 puts.add(new Callable<Void>() {
                     @Override
                     public Void call() throws Exception {
-                        app.sessionPut(key, value, session, false);
+                        app.sessionPut(key + "&sync=true", value, session, false);
                         return null;
                     }
                 });
@@ -175,9 +175,9 @@ public class SessionCacheOneServerTest extends FATServletClient {
             expectedValues.put(key, sb.toString());
         }
 
-        String sessionId = app.sessionPut("testConcurrentReplaceAttributes-key1", 100, session, true);
+        String sessionId = app.sessionPut("testConcurrentReplaceAttributes-key1&sync=true", 100, session, true);
         try {
-            app.sessionPut("testConcurrentReplaceAttributes-key2", 200, session, false);
+            app.sessionPut("testConcurrentReplaceAttributes-key2&sync=true", 200, session, false);
 
             List<Future<Void>> futures = executor.invokeAll(puts);
             for (Future<Void> future : futures)
@@ -228,7 +228,7 @@ public class SessionCacheOneServerTest extends FATServletClient {
                     requests.add(new Callable<Void>() {
                         @Override
                         public Void call() throws Exception {
-                            app.sessionPut("testConcurrentSetGetAndRemove-key", value, session, false);
+                            app.sessionPut("testConcurrentSetGetAndRemove-key&sync=true", value, session, false);
                             return null;
                         }
                     });
@@ -246,7 +246,7 @@ public class SessionCacheOneServerTest extends FATServletClient {
                     requests.add(new Callable<Void>() {
                         @Override
                         public Void call() throws Exception {
-                            app.invokeServlet("sessionRemoveAttribute&key=testConcurrentSetGetAndRemove-key", session);
+                            app.invokeServlet("sessionRemoveAttribute&key=testConcurrentSetGetAndRemove-key&sync=true", session);
                             return null;
                         }
                     });

--- a/dev/com.ibm.ws.session.cache_fat/test-applications/sessionCacheApp/src/session/cache/web/SessionCacheTestServlet.java
+++ b/dev/com.ibm.ws.session.cache_fat/test-applications/sessionCacheApp/src/session/cache/web/SessionCacheTestServlet.java
@@ -654,6 +654,14 @@ public class SessionCacheTestServlet extends FATServlet {
         String sessionID = session.getId();
         System.out.println("Put entry: " + key + '=' + value + " into sessionID=" + sessionID);
         response.getWriter().write("session id: [" + sessionID + "]");
+
+        // Normally, session postinvoke writes the updates after the servlet returns control to the test logic.
+        // This can be a problem if the test logic proceeds to execute further test logic based on the expectation
+        // that updates made under the previous servlet request have gone into effect.  Tests that are vulnerable
+        // to this can use the sync=true parameter to force update to be made before servlet returns.
+        boolean sync = Boolean.parseBoolean(request.getParameter("sync"));
+        if (sync)
+            ((IBMSession) session).sync();
     }
 
     public void sessionGet(HttpServletRequest request, HttpServletResponse response) throws Throwable {
@@ -683,6 +691,14 @@ public class SessionCacheTestServlet extends FATServlet {
         String key = request.getParameter("key");
         HttpSession session = request.getSession(false);
         session.removeAttribute(key);
+
+        // Normally, session postinvoke writes the updates after the servlet returns control to the test logic.
+        // This can be a problem if the test logic proceeds to execute further test logic based on the expectation
+        // that updates made under the previous servlet request have gone into effect.  Tests that are vulnerable
+        // to this can use the sync=true parameter to force update to be made before servlet returns.
+        boolean sync = Boolean.parseBoolean(request.getParameter("sync"));
+        if (sync)
+            ((IBMSession) session).sync();
     }
 
     public void sessionGetTimeout(HttpServletRequest request, HttpServletResponse response) throws Throwable {

